db/active_replicator_common.go

```diff
@@ -165,7 +165,6 @@ func (arc *activeReplicatorCommon) Start(ctx context.Context) error {
 			defer arc.ctxCancel()
 		} else {
 			base.InfofCtx(arc.ctx, base.KeyReplicate, "Attempting to reconnect in background: %v", err)
-			arc.reconnectActive.Set(true)
 			go arc.reconnectLoop()
```

Comment:
I feel like this change may potentially cause other races, where `reconnectActive.Set(true)` was previously being called inside the `arc.Lock`, and if we move it down into the `reconnectLoop` goroutine, there's a time at which there's an unlocked replicator that is about to go into a reconnecting state but hasn't yet set the flag?

Maybe races of this kind are guarded by other things, but I'm not as confident in this change as adding the `reconnectActive.Set` in the `OnExitCallback`. That seems like an obvious omission to me that doesn't affect any locking mechanisms.

Let me know if you disagree. I know this approach is easier to forget instances of `reconnectActive` for new `go arc.reconnectLoop()` invocations- but perhaps both lines could just be moved into a synchronous function?

Reply:
I think it would be fine to put the check inside the goroutine, I don't think there's a race to protect from. That said, I did spot a race which is in the reconnect function that it locks, then tries to connect. But if it is locked because `stop` was called (which acquires the lock), then the reconnect loop could reconnect after stop was called. Added a context cancel check again right after the lock to protect against it.

I think this refactor splits the difference between what we were going for.

Also, renamed misnamed `reconnect` function to `disconnect` in the refactor when I renamed `reconnectLoop` to `reconnect`.